### PR TITLE
Disallow Span and Memory from CA1836

### PR DESCRIPTION
**Pull request by Jozkee**
Fixes https://github.com/dotnet/roslyn-analyzers/issues/3810

cc @jeffhandley, @sharwell.

### CODE REVIEW

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs [hunk 1]
```diff
@@ -435,6 +435,22 @@ public Task CSharpTestWhitespaceTrivia()
         ? 0 :
         _dictionary.Count;
 }");
+
+        [Theory]
+        [InlineData("System.ReadOnlyMemory")]
+        [InlineData("System.ReadOnlySpan")]
+        [InlineData("System.Memory")]
+        [InlineData("System.Span")]
+        public Task CSharpTest_DisallowedTypesForCA1836_NoDiagnosis(string type)
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs:444]: [minor] Feel free to ignore but it seems that "most" tests seem to use the `_NoDiagnostic` suffix instead.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs [hunk 2]
```diff
@@ -435,6 +435,22 @@ public Task CSharpTestWhitespaceTrivia()
...
+            => VerifyCS.VerifyAnalyzerAsync(
```
Comment (Evangelink) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs:445]: Is it worth having the same test for vbnet?

Comment (Jozkee) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs:445]: These types are not supported on VB (yet?).
https://github.com/dotnet/vblang/issues/297

Comment (Evangelink) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs:445]: Oh! I thought that types were always available for all languages of the .net stack.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs (hunk 1 above)
Comment (Jozkee) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs:444]: This file in particular uses `NoDiagnosis` for all other tests so for now I will keep it as is.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCountProperly.cs [hunk 3]
```diff
@@ -127,50 +127,82 @@ private static void OnCompilationStart(CompilationStartAnalysisContext context)
...
-            static void AddIfNotNull(ImmutableHashSet<IMethodSymbol>.Builder set, IEnumerable<IMethodSymbol>? others)
+            static void AddIfNotNullMethodSymbols(ImmutableHashSet<IMethodSymbol>.Builder set, IEnumerable<IMethodSymbol>? others)
             {
                 if (others != null)
                 {
                     set.UnionWith(others);
                 }
             }
+
+            static void AddIfNotNullTypeSymbol(ImmutableHashSet<ITypeSymbol>.Builder set, ITypeSymbol? symbol)
```
[the remark this answers is not in the document]
Comment (Jozkee) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCountProperly.cs:192]: Will do, thanks.

[the remark this answers is not in the document]
Comment (Jozkee) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCountProperly.cs:192]: Nvm, I can't. `AddIfNotNullMethodSymbols` does `UnionWith` and this one does `Add`.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs (hunk 1 above)
Comment (Evangelink) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferIsEmptyOverCountTests.cs:444]: Oh sorry about that.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCountProperly.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCountProperly.cs:192]: Sorry looks like I wasn't clear enough, could you remove `AddIfNotNullTypeSymbol` and use the extension method `AddIfNotNull` instead? See https://github.com/dotnet/roslyn-analyzers/blob/194d1443c4f5901afd4142587f88c35b06b37654/src/Utilities/Compiler/Extensions/ImmutableHashSetExtensions.cs#L95-L102

[the remark this answers is not in the document]
Comment (Jozkee) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseCountProperly.cs]: Done.

**Comment by Jozkee** [comment]
Follow-up items:
* Update documentation to reflect the exclusion. The code sample was using one of the disallowed types.
* Change severity of the rule to Warning in dotnet/runtime 
https://github.com/dotnet/runtime/blob/0a1f7ba9475c78e6a8894a076f57a05baa6c1697/eng/CodeAnalysis.ruleset#L114
* Undo changes in https://github.com/dotnet/runtime/pull/37529 for the disallowed types.

**Pull request merged**